tests/test_radiation.py

```diff
@@ -0,0 +1,212 @@
+    argnames="input, exp_tau, exp_ppfd",
+    argvalues=[
+        (29376000, 0.752, 43.713),
+        (
+            np.array([29376000, 29376000]),
+            np.array([0.752, 0.752]),
+            np.array([43.713, 43.713]),
+        ),
+    ],
+)
+def test_calc_ppfd(radiation_fixture, input, exp_tau, exp_ppfd):
+    """Simple check for correct numbers, better test to be decided."""
+
+    tau = radiation_fixture.calc_ppfd(input, 1.0)
+    assert tau == pytest.approx(exp_tau, 0.1)
+    assert radiation_fixture.ppfd == pytest.approx(exp_ppfd, 0.1)
+
+
+@pytest.mark.parametrize(
+    argnames="input, exp_tau, exp_toc_radiation",
+    argvalues=[
+        (29376000, 0.752, 18335385.16),
+        (
+            np.array([29376000, 29376000]),
+            np.array([0.752, 0.752]),
+            np.array([18335385.16, 18335385.16]),
+        ),
+    ],
+)
+def test_calc_topofcanopy_radiation(
+    radiation_fixture, input, exp_tau, exp_toc_radiation
+):
+    """Simple check for correct numbers, better test to be decided."""
+
+    tau = radiation_fixture.calc_ppfd(input, 1.0)
+    assert tau == pytest.approx(exp_tau, 0.1)
+    radiation_fixture.calc_topofcanopy_radiation(tau, input, 1.0)
+    assert radiation_fixture.topofcanopy_radiation == pytest.approx(
+        exp_toc_radiation, 0.1
+    )
+
+
+@pytest.mark.parametrize(
+    argnames="temp_canopy, temp_soil, exp_lw_canopy, exp_lw_soil",
+    argvalues=[
+        (25, 25, 425.6434, 425.6434),
+        (
+            np.array([25, 25]),
+            np.array([25, 25]),
+            np.array([425.6434, 425.6434]),
+            np.array([425.6434, 425.6434]),
+        ),
+    ],
+)
+def test_calc_longwave_radiation(
+    radiation_fixture, temp_canopy, temp_soil, exp_lw_canopy, exp_lw_soil
+):
+    """Simple check for correct numbers, better test to be decided."""
+
+    radiation_fixture.calc_longwave_radiation(temp_canopy, temp_soil)
+    assert radiation_fixture.longwave_canopy == pytest.approx(exp_lw_canopy)
+    assert radiation_fixture.longwave_soil == pytest.approx(exp_lw_soil)
+
+
+@pytest.mark.parametrize(
+    argnames="toc_radiation, lw_canopy, lw_soil, rad_absorbed, exp_netrad",
+    argvalues=[
+        (10000, 100, 100, 200, 9600),
+        (
+            np.array([10000, 10000]),
+            np.array([100, 100]),
+            np.array([100, 100]),
+            np.array([200, 200]),
+            np.array([9600, 9600]),
+        ),
+    ],
+)
+def test_calc_netradiation_surface(
+    radiation_fixture, toc_radiation, lw_canopy, lw_soil, rad_absorbed, exp_netrad
+):
+    """Simple check for correct numbers, better test to be decided."""
+
+    radiation_fixture.topofcanopy_radiation = toc_radiation
+    radiation_fixture.longwave_canopy = lw_canopy
+    radiation_fixture.longwave_soil = lw_soil
+    radiation_fixture.calc_netradiation_surface(rad_absorbed)
+    assert radiation_fixture.netradiation_surface == pytest.approx(exp_netrad)
+
+
+@pytest.mark.parametrize(
```

Comment:
I was told on one of my pull requests to add a wider range of values to unit tests, so think you probably should here. As what your `parameterise` is testing here is basically that the functions work for a single float and an array of (the same) float.

Reply:
Obviously depends a bit on how pinned down the parameterisation is, not work sinking loads of time into developing extensive tests for functions that don't have remotely sensible parameters

Reply:
It's useful to have a few values covering maybe minimum, median and maximum sensible values? Honestly with float functions you are very unlikely to have one test value stay unchanged when an error is introduced - although it could be at some kind of stable point or limit* - but having a few is a bit safer from the point of view of having test values that act as a canary when the changing the code changes predictions unexpectedly.

* I'm just basically selecting maths words at random here.

Reply:
Good point, it makes sense to include a wider range of values including some non-valid ones to check that it fails. Would it be worth using an advance method for that or could I just extend the parameterize functionality I currently use?

Reply:
I would probably just extend the parameterisation. This [existing test](https://github.com/ImperialCollegeLondon/virtual_rainforest/blob/develop/tests/test_data.py#L305) provides a possible template for adding failing values in - you can add an expected Exception (or expected lack of exception) to the parameter sets. 

I really like the `pytest.param` and then `id=param_set_name` syntax used in that example - the name acts both as a short explanation of what those inputs test and also makes it easier to identify failing tests.

Reply:
The steps need separating:

```python
@pytest.mark.parametrize(
    argnames="elev, sun_frac, exp_tau, exp_err",
    argvalues=[
        pytest.param(
            np.array([100, 1000]),
            np.array([0.5, 1.0]),
            np.array([0.501335, 0.770025]),
            does_not_raise(),
            id="standard_array_should_get",
        ),
        pytest.param(
            np.array([100, 1000]),
            np.array([1.1, 1.0]),
            None,
            pytest.raises(ValueError),
            id="value_error_sf",
        ),
    ],
)
def test_calculate_atmospheric_transmissivity(elev, sun_frac, exp_tau, exp_err):
    """Test atmospheric tansmissivity across elevation and sunshine fractions."""

    # Run the code that is expected to raise the exception, testing the assertion that it is raised
    with exp_err:
        result = radiation.calculate_atmospheric_transmissivity(elev, sun_frac)
        # If the code does not raise, we can now proceed to testing the result
        assert np.allclose(result, exp_tau)
```

Reply:
Note that the `assert np.allclose(result, exp_tau)` has to happen within the `with` block, because otherwise the `assert` will be tested even if the code raises an exception.

Reply:
hmm it still doesn't like the `np.allclose` for `None`

Reply:
Oh - thought that would work. HOw about this:

```python
  # Run the code that is expected to raise the exception, testing the assertion that it is raised
  with exp_err as err:
      result = radiation.calculate_atmospheric_transmissivity(elev, sun_frac)

if not err:
      assert np.allclose(result, exp_tau)
```

Reply:
Perfect, this works, thank you! Now I just need to think of ways in which this could go wrong ;-)